# Future Sales Prediction - Kaggle Competition

### Modeling File

https://www.kaggle.com/competitions/competitive-data-science-predict-future-sales/

#### File descriptions
- sales_train.csv - the training set. Daily historical data from January 2013 to October 2015.

- test.csv - the test set. You need to forecast the sales for these shops and products for November 2015.

- sample_submission.csv - a sample submission file in the correct format.

- items.csv - supplemental information about the items/products.

- item_categories.csv  - supplemental information about the items categories.

- shops.csv- supplemental information about the shops.

#### Data fields

- ID - an Id that represents a (Shop, Item) tuple within the test set

- shop_id - unique identifier of a shop

- item_id - unique identifier of a product

- item_category_id - unique identifier of item category

- item_cnt_day - number of products sold. You are predicting a monthly amount of this measure

- item_price - current price of an item

- date - date in format dd/mm/yyyy

- date_block_num - a consecutive month number, used for convenience. January 2013 is 0, February 2013 is 1,..., October 2015 is 33

- item_name - name of item

- shop_name - name of shop

- item_category_name - name of item category

### I. Import the dataset and data analysis libraries

In [85]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [86]:
sales_train = pd.read_csv("data/sales_train.csv")
sales_train.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day
0,02.01.2013,0,59,22154,999.00,1.0
1,03.01.2013,0,25,2552,899.00,1.0
2,05.01.2013,0,25,2552,899.00,-1.0
3,06.01.2013,0,25,2554,1709.05,1.0
4,15.01.2013,0,25,2555,1099.00,1.0


In [87]:
test = pd.read_csv("data/test.csv")
test.head()

,ID,shop_id,item_id
0,0,5,5037
1,1,5,5320
2,2,5,5233
3,3,5,5232
4,4,5,5268


In [88]:
sample_submission = pd.read_csv("data/sample_submission.csv")
print(sample_submission.shape)
sample_submission.head()

(214200, 2)


,ID,item_cnt_month
0,0,0.5
1,1,0.5
2,2,0.5
3,3,0.5
4,4,0.5


In [89]:
items = pd.read_csv("data/items.csv")
items.head()

,item_name,item_id,item_category_id
0,! ВО ВЛАСТИ НАВАЖДЕНИЯ (ПЛАСТ.) D,0,40
1,!ABBYY FineReader 12 Professional Edition Full...,1,76
2,***В ЛУЧАХ СЛАВЫ (UNV) D,2,40
3,***ГОЛУБАЯ ВОЛНА (Univ) D,3,40
4,***КОРОБКА (СТЕКЛО) D,4,40


In [90]:
item_categories = pd.read_csv("data/item_categories.csv")
item_categories.head()

,item_category_name,item_category_id
0,PC - Гарнитуры/Наушники,0
1,Аксессуары - PS2,1
2,Аксессуары - PS3,2
3,Аксессуары - PS4,3
4,Аксессуары - PSP,4


In [91]:
shops = pd.read_csv("data/shops.csv")
shops.head()

,shop_name,shop_id
0,"!Якутск Орджоникидзе, 56 фран",0
1,"!Якутск ТЦ ""Центральный"" фран",1
2,"Адыгея ТЦ ""Мега""",2
3,"Балашиха ТРК ""Октябрь-Киномир""",3
4,"Волжский ТЦ ""Волга Молл""",4


### II. Initial data checks and manipulations

In [92]:
sales_train['date'] = pd.to_datetime(sales_train['date'], format = '%d.%m.%Y')
sales_train.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day
0,2013-01-02,0,59,22154,999.00,1.0
1,2013-01-03,0,25,2552,899.00,1.0
2,2013-01-05,0,25,2552,899.00,-1.0
3,2013-01-06,0,25,2554,1709.05,1.0
4,2013-01-15,0,25,2555,1099.00,1.0


In [93]:
# Split into month, day, and year
sales_train['year'] = sales_train['date'].dt.year
sales_train['month'] = sales_train['date'].dt.month
sales_train['day'] = sales_train['date'].dt.day
sales_train.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day,year,month,day
0,2013-01-02,0,59,22154,999.00,1.0,2013,1,2
1,2013-01-03,0,25,2552,899.00,1.0,2013,1,3
2,2013-01-05,0,25,2552,899.00,-1.0,2013,1,5
3,2013-01-06,0,25,2554,1709.05,1.0,2013,1,6
4,2013-01-15,0,25,2555,1099.00,1.0,2013,1,15


In [94]:
sales_train.isna().sum()

date              0
date_block_num    0
shop_id           0
item_id           0
item_price        0
item_cnt_day      0
year              0
month             0
day               0
dtype: int64

In [95]:
sales_train.nunique()

date               1034
date_block_num       34
shop_id              60
item_id           21807
item_price        19993
item_cnt_day        198
year                  3
month                12
day                  31
dtype: int64

In [96]:
# Describing datetime and quantitative variables
sales_train[['date', 'item_price', 'item_cnt_day']].describe()

,date,item_price,item_cnt_day
count,2935849,2.935849e+06,2.935849e+06
mean,2014-04-03 05:44:34.970681344,8.908532e+02,1.242641e+00
min,2013-01-01 00:00:00,-1.000000e+00,-2.200000e+01
25%,2013-08-01 00:00:00,2.490000e+02,1.000000e+00
50%,2014-03-04 00:00:00,3.990000e+02,1.000000e+00
75%,2014-12-05 00:00:00,9.990000e+02,1.000000e+00
max,2015-10-31 00:00:00,3.079800e+05,2.169000e+03
std,NaN,1.729800e+03,2.618834e+00


In [97]:
# Declaring categorical variables
sales_train['shop_id'] = pd.Categorical(sales_train['shop_id'])
sales_train['item_id'] = pd.Categorical(sales_train['item_id'])

In [98]:
# Convert negative to positive values in item_cnt_day
sales_train['item_cnt_day'] = sales_train['item_cnt_day'].abs()

In [99]:
# Create a "master table" for EDA in training data
sales_full = sales_train.merge(items, how = 'inner', on = 'item_id')\
                              .merge(item_categories, how = 'inner', on = 'item_category_id')\
                              .merge(shops, how = 'inner', on = 'shop_id')

sales_full.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day,year,month,day,item_name,item_category_id,item_category_name,shop_name
0,2013-01-02,0,59,22154,999.00,1.0,2013,1,2,ЯВЛЕНИЕ 2012 (BD),37,Кино - Blu-Ray,"Ярославль ТЦ ""Альтаир"""
1,2013-01-03,0,25,2552,899.00,1.0,2013,1,3,DEEP PURPLE The House Of Blue Light LP,58,Музыка - Винил,"Москва ТРК ""Атриум"""
2,2013-01-05,0,25,2552,899.00,1.0,2013,1,5,DEEP PURPLE The House Of Blue Light LP,58,Музыка - Винил,"Москва ТРК ""Атриум"""
3,2013-01-06,0,25,2554,1709.05,1.0,2013,1,6,DEEP PURPLE Who Do You Think We Are LP,58,Музыка - Винил,"Москва ТРК ""Атриум"""
4,2013-01-15,0,25,2555,1099.00,1.0,2013,1,15,DEEP PURPLE 30 Very Best Of 2CD (Фирм.),56,Музыка - CD фирменного производства,"Москва ТРК ""Атриум"""


#### Data checks:

1. Is there any item sold at more than one different price?

In [100]:
shops_and_items = sales_full[['shop_id', 'item_id', 'item_price']].copy()
shops_and_items.drop_duplicates(['shop_id', 'item_id'], inplace = True)

# Same shop, same item, different prices?
print(len(shops_and_items[shops_and_items.duplicated(subset = ['shop_id', 'item_id', 'item_price'])]))

0


2. Is there any group of items having different IDs but the same name? Do the same with shop name and item category name.

In [101]:
print(items.duplicated(['item_id']).sum()) # Number of duplicated item IDs
print(items.duplicated(['item_name']).sum()) # Number of duplicated item names

0
0


In [102]:
# With item categories
print(item_categories.duplicated(['item_category_id']).sum())
print(item_categories.duplicated(['item_category_name']).sum())

0
0


In [103]:
# With shops
print(shops.duplicated(['shop_id']).sum())
print(shops.duplicated(['shop_name']).sum())

0
0


3. Is there any day when the shop can't sell any item?

In [104]:
# Generate list of datetime indices - all days of sales in the training data
datetime_indices = sales_full[['date', 'date_block_num']].copy()
datetime_indices.drop_duplicates(inplace = True)
datetime_indices.sort_values('date', inplace = True)
datetime_indices.set_index('date', inplace = True)
print(datetime_indices.index)

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
               '2013-01-09', '2013-01-10',
               ...
               '2015-10-22', '2015-10-23', '2015-10-24', '2015-10-25',
               '2015-10-26', '2015-10-27', '2015-10-28', '2015-10-29',
               '2015-10-30', '2015-10-31'],
              dtype='datetime64[ns]', name='date', length=1034, freq=None)


In [105]:
# Generate a list of dates from the first and last day of sales in the training data
date_range = pd.date_range(start = sales_full['date'].min(), end = sales_full['date'].max())
print(date_range)

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
               '2013-01-09', '2013-01-10',
               ...
               '2015-10-22', '2015-10-23', '2015-10-24', '2015-10-25',
               '2015-10-26', '2015-10-27', '2015-10-28', '2015-10-29',
               '2015-10-30', '2015-10-31'],
              dtype='datetime64[ns]', length=1034, freq='D')


#### Fuzzy string matching - a part of data check:

Are there any pair of item names, shop names, or item category names close to each other? If this is possible, it might be the case of spelling or processing errors, or two categories which can be grouped to one. (Kaggle community said probably)

In [106]:
# https://typesense.org/learn/fuzzy-string-matching-python/
# https://pypi.org/project/fuzzywuzzy/
# https://towardsdatascience.com/a-fuzzy-string-matching-story-314bbecaa098/

from itertools import combinations, chain
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

In [107]:
def fuzzy_string_matching_metrics(df, name_column):
    '''df: the DataFrame in question
    name_column: the column name that you want to check for fuzzy string matching (must be a string)'''

    name_list = df[name_column].tolist() # extract name pairs and convert them to a Python list
    pairwise_combinations = list(combinations(name_list, 2)) # generate all pairwise combinations of shop names to calculate similarity

    # https://stackoverflow.com/questions/7422453/python-change-type-of-whole-list
    # Change each element of the pairwise combinations into a list, then create a DataFrame for easier sorting
    pairwise_combinations = list(map(list, pairwise_combinations))
    pairwise_df = pd.DataFrame(pairwise_combinations, columns = ['first_string', 'second_string'])

    # Compute various similarity metrics
    pairwise_df['fuzz_ratio'] = pairwise_df.apply(lambda row: fuzz.ratio(row['first_string'], row['second_string']), axis = 1)
    pairwise_df['partial_ratio'] = pairwise_df.apply(lambda row: fuzz.partial_ratio(row['first_string'], row['second_string']), axis = 1)
    pairwise_df['token_sort_ratio'] = pairwise_df.apply(lambda row: fuzz.token_sort_ratio(row['first_string'], row['second_string']), axis = 1)
    pairwise_df['token_set_ratio'] = pairwise_df.apply(lambda row: fuzz.token_set_ratio(row['first_string'], row['second_string']), axis = 1)

    return pairwise_df

In [108]:
def sort_desc_with_threshold(df, sort_column, threshold):
    # Keep the necessary columns
    df_filtered = df[['first_string', 'second_string', sort_column]]

    # Filter for values greater or equal to a threshold 
    df_filtered = df_filtered[df_filtered[sort_column] >= threshold]

    # Sort values in descending order
    df_filtered.sort_values(sort_column, ascending = False, inplace = True)
    return df_filtered

In [109]:
def get_unique_values(*args):
    '''Pass individual list of lists as inputs. Returns a list of lists.'''

    # Combine the list of lists together and flatten it to be a 2D list
    whole_list = list(chain(args))
    flattened_list = [item for sublist in whole_list for item in sublist]

    # Convert each element of individual list to tuple format to later use the "set" data type (keeping unique values only)
    list_of_tuples = [tuple(x) for x in flattened_list]
    unique_val_set = set(list_of_tuples) # return a set object

    # Convert back into list of tuples and then list of lists
    unique_val_list = list(unique_val_set)
    unique_val_list = [list(x) for x in unique_val_list]

    return unique_val_list

In [110]:
def flag_similar_pairs(original_df, name_column, fuzz_ratio_threshold = 90, partial_ratio_threshold = 90,
                       token_sort_ratio_threshold = 90, token_set_ratio_threshold = 90):
    '''You can change the similarity thresholds by modifying the default arguments in this function.'''
    
    # Generate a DataFrame with fuzzy string matching metrics of all pairs of names in a particular DataFrame column
    df = fuzzy_string_matching_metrics(original_df, name_column)

    # Find similar pairs with high fuzz ratio, partial ratio, token sort ratio and token set ratio
    top_fuzz_ratios = sort_desc_with_threshold(df, 'fuzz_ratio', fuzz_ratio_threshold)[['first_string', 'second_string']].values.tolist()
    top_partial_ratios = sort_desc_with_threshold(df, 'partial_ratio', partial_ratio_threshold)[['first_string', 'second_string']].values.tolist()
    top_token_sort_ratios = sort_desc_with_threshold(df, 'token_sort_ratio', token_sort_ratio_threshold)[['first_string', 'second_string']].values.tolist()
    top_token_set_ratios = sort_desc_with_threshold(df, 'token_set_ratio', token_set_ratio_threshold)[['first_string', 'second_string']].values.tolist()

    # Get unique values that are high in at least one metric...
    unique_pairs_list = get_unique_values(top_fuzz_ratios, top_partial_ratios, top_token_sort_ratios, top_token_set_ratios)
    # ...then output a DataFrame for better readability
    result_df = df[df[['first_string', 'second_string']].apply(list, axis = 1).isin(unique_pairs_list)]
    return result_df

**For shop names:**

Because I have written so many functions, I decided to show the intermediate steps to arrive at finding similar pairs of names for this particular example.

In [111]:
shops.head()

,shop_name,shop_id
0,"!Якутск Орджоникидзе, 56 фран",0
1,"!Якутск ТЦ ""Центральный"" фран",1
2,"Адыгея ТЦ ""Мега""",2
3,"Балашиха ТРК ""Октябрь-Киномир""",3
4,"Волжский ТЦ ""Волга Молл""",4


In [112]:
# Example of a table of fuzzy string matching metrics
fuzzy_shop_names = fuzzy_string_matching_metrics(shops, 'shop_name')
fuzzy_shop_names.head()

,first_string,second_string,fuzz_ratio,partial_ratio,token_sort_ratio,token_set_ratio
0,"!Якутск Орджоникидзе, 56 фран","!Якутск ТЦ ""Центральный"" фран",52,52,53,59
1,"!Якутск Орджоникидзе, 56 фран","Адыгея ТЦ ""Мега""",22,33,29,29
2,"!Якутск Орджоникидзе, 56 фран","Балашиха ТРК ""Октябрь-Киномир""",20,21,33,33
3,"!Якутск Орджоникидзе, 56 фран","Волжский ТЦ ""Волга Молл""",23,25,33,33
4,"!Якутск Орджоникидзе, 56 фран","Вологда ТРЦ ""Мармелад""",24,28,34,34


In [113]:
# Example of keeping pairs with the highest fuzz ratio and sorting them in descending order
top_fuzz_ratio = sort_desc_with_threshold(fuzzy_shop_names, 'fuzz_ratio', 80)
top_fuzz_ratio

,first_string,second_string,fuzz_ratio
545,Жуковский ул. Чкалова 39м?,Жуковский ул. Чкалова 39м²,96
1104,"Москва ТК ""Буденовский"" (пав.А2)","Москва ТК ""Буденовский"" (пав.К7)",94
56,"!Якутск Орджоникидзе, 56 фран","Якутск Орджоникидзе, 56",88
115,"!Якутск ТЦ ""Центральный"" фран","Якутск ТЦ ""Центральный""",88
1560,"РостовНаДону ТРК ""Мегацентр Горизонт""","РостовНаДону ТРК ""Мегацентр Горизонт"" Островной",88
1335,"Москва ТЦ ""Перловский""","Москва ТЦ ""Семеновский""",84
1554,"Омск ТЦ ""Мега""","Химки ТЦ ""Мега""",83


In [114]:
# When the whole function is completed...
flag_similar_pairs(shops, 'shop_name')

,first_string,second_string,fuzz_ratio,partial_ratio,token_sort_ratio,token_set_ratio
56,"!Якутск Орджоникидзе, 56 фран","Якутск Орджоникидзе, 56",88,100,90,100
115,"!Якутск ТЦ ""Центральный"" фран","Якутск ТЦ ""Центральный""",88,100,89,100
545,Жуковский ул. Чкалова 39м?,Жуковский ул. Чкалова 39м²,96,96,100,100
1104,"Москва ТК ""Буденовский"" (пав.А2)","Москва ТК ""Буденовский"" (пав.К7)",94,94,89,94
1560,"РостовНаДону ТРК ""Мегацентр Горизонт""","РостовНаДону ТРК ""Мегацентр Горизонт"" Островной",88,100,88,100
1561,"РостовНаДону ТРК ""Мегацентр Горизонт""","РостовНаДону ТЦ ""Мега""",71,91,69,75
1580,"РостовНаДону ТРК ""Мегацентр Горизонт"" Островной","РостовНаДону ТЦ ""Мега""",61,91,58,75


There are some particularly interesting similar name combinations that have a high similarity (at least a similarity metric >= 90/100). I will use Google Translate to try to understand them.

- Index 56: ```!Якутск Орджоникидзе, 56 фран``` (!Yakutsk Ordzhonikidze, 56 fran) and ```Якутск Орджоникидзе, 56``` (Yakutsk Ordzhonikidze).

- Index 115: ```!Якутск ТЦ "Центральный" фран``` (!Yakutsk shopping center "Central" franc) and ```Якутск ТЦ "Центральный"``` (Yakutsk shopping center "Central").

- Index 545: ```Жуковский ул. Чкалова 39м?``` (Zhukovsky st. Chkalova 39m?) and ```Жуковский ул. Чкалова 39м²``` (Zhukovsky st. Chkalova 39m²).

- Index 1104: ```Москва ТК "Буденовский" (пав.А2)``` (Moscow TC "Budenovsky" (pavilion A2)) and ```Москва ТК "Буденовский" (пав.К7)``` (Moscow TC "Budenovsky" (pavement K7)).

- Index 1560, 1561 and 1580: ```РостовНаДону ТРК "Мегацентр Горизонт"``` (Rostov-on-Don TRC "Megacenter Gorizont"), ```РостовНаДону ТРК "Мегацентр Горизонт" Островной``` (Rostov-on-Don TRC "Megacenter Gorizont" Island) and ```РостовНаДону ТЦ "Мега"``` (Rostov-on-Don TC "Mega").

**Fuzzy string matching results:**

- ```фран``` (56, 115) does not have a clear meaning in Russian, so I think that this is a result of inaccurate processing or data entry.

- There are some unwanted special characters, such as ```!``` and ```?``` (56, 115, 545), that made spotting duplicated shop names impossible without fuzzy string matching.

- I have asked ChatGPT to save my time browsing through Russian sources while I don't understand anything about it, and it said that while ```РостовНаДону ТРК "Мегацентр Горизонт"``` and ```РостовНаДону ТРК "Мегацентр Горизонт" Островной``` (1560) refer to the same shop, ```РостовНаДону ТЦ "Мега"``` is another shopping center separated from these two names.

**For item category names:**

In [115]:
item_categories.head()

,item_category_name,item_category_id
0,PC - Гарнитуры/Наушники,0
1,Аксессуары - PS2,1
2,Аксессуары - PS3,2
3,Аксессуары - PS4,3
4,Аксессуары - PSP,4


In [116]:
flag_similar_pairs(item_categories, 'item_category_name')

,first_string,second_string,fuzz_ratio,partial_ratio,token_sort_ratio,token_set_ratio
83,Аксессуары - PS2,Аксессуары - PS3,94,94,93,93
84,Аксессуары - PS2,Аксессуары - PS4,94,94,93,93
85,Аксессуары - PS2,Аксессуары - PSP,94,94,93,93
86,Аксессуары - PS2,Аксессуары - PSVita,86,94,84,84
165,Аксессуары - PS3,Аксессуары - PS4,94,94,93,93
166,Аксессуары - PS3,Аксессуары - PSP,94,94,93,93
167,Аксессуары - PS3,Аксессуары - PSVita,86,94,84,84
246,Аксессуары - PS4,Аксессуары - PSP,94,94,93,93
247,Аксессуары - PS4,Аксессуары - PSVita,86,94,84,84
326,Аксессуары - PSP,Аксессуары - PSVita,86,94,84,84


**Decoding similar category names with Google Translate:**

| Russian      | English       | Russian   | English | Russian         | English       |
|--------------|---------------|-----------|---------|-----------------|---------------|
| Аксессуары   | Accessories   | Книги     | Books   | Игровые консоли | Game consoles |
| Подарки      | Gifts         | Игры      | Games   | Программы       | Programs      |
| Карты оплаты | Payment cards | Служебные | Utility | Кино            | Movies        |

**Special word combinations in Russian:**

- ```Книги - Аудиокниги (Цифра)```: Books - Audiobooks (Digital)

- ```Подарки - Настольные игры (компактные)```: Gifts - Board games (compact)

- ```Подарки - Сувениры (в навеску)```: Gifts - Souvenirs (as a gift)

- ```Программы - Для дома и офиса (Цифра)```: Programs - For home and office (Digital)

- ```Программы - Обучающие (Цифра)```: Programs - Educational (Digital)

- ```Служебные - Билеты```: Service - Tickets

**String similarity matching result:**

- Although being different types, ```Аксессуары``` (accessories), ```Игровые консоли``` (game consoles), and ```Игры``` (games) still have categories with the same names such as PS2, PS3, or PS4. For feature engineering, I can regroup each PS product into an item category.

- For the annotations (in the **Special word combinations** part):

    + ```Цифра``` (digital), ```компактные``` (compact), and ```Билеты``` (tickets) are additional information to distinguish between different item categories. I have decided to keep them as a separate category.

    + Because ```Подарки``` is already "gift" in Russian, the additional annotation ```в навеску``` (as a gift) is only extra information. I will merge ```Подарки - Сувениры``` and ```Подарки - Сувениры (в навеску)``` into a single category.

Python outputs ```MemoryError``` when I was working with item names (>20,000 different names), but I think that cleaning the shop names and item category names is enough.

#### Applying fuzzy string matching results to the sales dataset:

In [117]:
# Stripping sales_full of item, shop, and item category names
sales_full.drop(['item_name', 'item_category_name', 'shop_name'], axis = 1, inplace = True)
sales_full.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day,year,month,day,item_category_id
0,2013-01-02,0,59,22154,999.00,1.0,2013,1,2,37
1,2013-01-03,0,25,2552,899.00,1.0,2013,1,3,58
2,2013-01-05,0,25,2552,899.00,1.0,2013,1,5,58
3,2013-01-06,0,25,2554,1709.05,1.0,2013,1,6,58
4,2013-01-15,0,25,2555,1099.00,1.0,2013,1,15,56


**Cleaning shop data:**

In [118]:
shops[shops['shop_name'].isin(['!Якутск Орджоникидзе, 56 фран', 'Якутск Орджоникидзе, 56'])]

,shop_name,shop_id
0,"!Якутск Орджоникидзе, 56 фран",0
57,"Якутск Орджоникидзе, 56",57


In [119]:
shops[shops['shop_name'].isin(['!Якутск ТЦ "Центральный" фран', 'Якутск ТЦ "Центральный"'])]

,shop_name,shop_id
1,"!Якутск ТЦ ""Центральный"" фран",1
58,"Якутск ТЦ ""Центральный""",58


In [120]:
shops[shops['shop_name'].isin(['Жуковский ул. Чкалова 39м?', 'Жуковский ул. Чкалова 39м²'])]

,shop_name,shop_id
10,Жуковский ул. Чкалова 39м?,10
11,Жуковский ул. Чкалова 39м²,11


In [121]:
shops[shops['shop_name'].isin(['РостовНаДону ТРК "Мегацентр Горизонт"', 'РостовНаДону ТРК "Мегацентр Горизонт" Островной'])]

,shop_name,shop_id
39,"РостовНаДону ТРК ""Мегацентр Горизонт""",39
40,"РостовНаДону ТРК ""Мегацентр Горизонт"" Островной",40


In [122]:
# Change the necessary shop IDs
sales_full['shop_id'] = sales_full['shop_id'].copy().replace(0, 57)
sales_full['shop_id'] = sales_full['shop_id'].copy().replace(1, 58)
sales_full['shop_id'] = sales_full['shop_id'].copy().replace(10, 11)
sales_full['shop_id'] = sales_full['shop_id'].copy().replace(40, 39)

sales_full.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day,year,month,day,item_category_id
0,2013-01-02,0,59,22154,999.00,1.0,2013,1,2,37
1,2013-01-03,0,25,2552,899.00,1.0,2013,1,3,58
2,2013-01-05,0,25,2552,899.00,1.0,2013,1,5,58
3,2013-01-06,0,25,2554,1709.05,1.0,2013,1,6,58
4,2013-01-15,0,25,2555,1099.00,1.0,2013,1,15,56


**Cleaning item category data:**

In [123]:
item_categories.head()

,item_category_name,item_category_id
0,PC - Гарнитуры/Наушники,0
1,Аксессуары - PS2,1
2,Аксессуары - PS3,2
3,Аксессуары - PS4,3
4,Аксессуары - PSP,4


In [124]:
item_categories[item_categories['item_category_name'].str.contains('PS')]

,item_category_name,item_category_id
1,Аксессуары - PS2,1
2,Аксессуары - PS3,2
3,Аксессуары - PS4,3
4,Аксессуары - PSP,4
5,Аксессуары - PSVita,5
10,Игровые консоли - PS2,10
11,Игровые консоли - PS3,11
12,Игровые консоли - PS4,12
13,Игровые консоли - PSP,13
14,Игровые консоли - PSVita,14


In [125]:
item_categories[item_categories['item_category_name'].str.contains('Подарки - Сувениры')]

,item_category_name,item_category_id
69,Подарки - Сувениры,69
70,Подарки - Сувениры (в навеску),70


In [126]:
# Change the necessary item category IDs
sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(10, 1) # for PS2
sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(18, 1)

sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(11, 2) # for PS3
sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(19, 2)

sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(12, 3) # for PS4
sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(20, 3)

sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(13, 4) # for PSP
sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(21, 4)

sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(14, 5) # for PSVita
sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(22, 5)

sales_full['item_category_id'] = sales_full['item_category_id'].copy().replace(70, 69) # for "Подарки - Сувениры"

sales_full.head()

,date,date_block_num,shop_id,item_id,item_price,item_cnt_day,year,month,day,item_category_id
0,2013-01-02,0,59,22154,999.00,1.0,2013,1,2,37
1,2013-01-03,0,25,2552,899.00,1.0,2013,1,3,58
2,2013-01-05,0,25,2552,899.00,1.0,2013,1,5,58
3,2013-01-06,0,25,2554,1709.05,1.0,2013,1,6,58
4,2013-01-15,0,25,2555,1099.00,1.0,2013,1,15,56


### III. Feature engineering

Here, we consider ```date_block_num``` variable as number of months elapsed after January 2013, and this column can also be useful for splitting into different sets.

Aggregate existing features:

In [127]:
# Keep the necessary columns only - we are predicting by month
sales_features = sales_full[['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_price', 'item_cnt_day']].copy()

# Because no item has more than 1 price, we can group item price by the median
sales_features_price = sales_features[['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_price']]\
    .groupby(['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id'], as_index = False).median()

# Group item count by month by the sum
sales_features_itemcnt = sales_features[['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_cnt_day']]\
    .groupby(['year', 'month', 'shop_id', 'item_category_id', 'item_id'], as_index = False).sum()

sales_features_itemcnt.rename(columns = {'item_cnt_day': 'item_cnt_month'}, inplace = True)

sales_features = sales_features_price.merge(sales_features_itemcnt, how = 'inner', on = ['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id'])
sales_features.sort_values(['year', 'month', 'date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_price', 'item_cnt_month'], inplace = True)
sales_features.head()

,year,month,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month
0,2013,1,0,2,2,27,2499.0,1.0
1,2013,1,0,2,2,1409,1398.5,1.0
2,2013,1,0,2,2,1467,899.0,1.0
3,2013,1,0,2,2,1471,2599.0,2.0
4,2013,1,0,2,2,1832,1999.0,1.0


Create a separate table to match ```date_block_num``` with months and years:

In [128]:
# Guide for date block number
date_block = sales_features[['year', 'month', 'date_block_num']].copy().drop_duplicates(ignore_index = True)
date_block.head()

,year,month,date_block_num
0,2013,1,0
1,2013,2,1
2,2013,3,2
3,2013,4,3
4,2013,5,4


#### 1. Number of unique items in each shop/category for each month and median price of each shop/category/item

In [129]:
# Number of unique items and median price in each shop

shops_and_items = sales_features[['year', 'month', 'shop_id', 'item_price', 'item_id']].copy()

shops_and_items = shops_and_items.groupby(['year', 'month', 'shop_id'], as_index = False)\
    .agg({'item_price': 'median', 'item_id': 'nunique'})
shops_and_items.rename(columns = {'item_price': 'med_price_shop', 'item_id': 'unique_items_in_shop'}, inplace = True)

shops_and_items.head()

,year,month,shop_id,med_price_shop,unique_items_in_shop
0,2013,1,2,399.0,728
1,2013,1,3,399.0,544
2,2013,1,4,299.0,1062
3,2013,1,6,349.0,1865
4,2013,1,7,349.0,1271


In [130]:
# Number of unique items and median price in each item category

cat_and_items = sales_features[['year', 'month', 'item_category_id', 'item_price', 'item_id']].copy()

cat_and_items = cat_and_items.groupby(['year', 'month', 'item_category_id'], as_index = False)\
    .agg({'item_price': 'median', 'item_id': 'nunique'})
cat_and_items.rename(columns = {'item_price': 'med_price_cat', 'item_id': 'unique_items_in_cat'}, inplace = True)

cat_and_items.head()

,year,month,item_category_id,med_price_cat,unique_items_in_cat
0,2013,1,0,148.0,1
1,2013,1,1,398.0,2
2,2013,1,2,1499.0,289
3,2013,1,3,499.0,2
4,2013,1,4,599.0,96


In [131]:
# Monthly median price of all items sold in the month in general

med_price_items = sales_features[['year', 'month', 'item_price']].copy()

med_price_items = med_price_items.groupby(['year', 'month'], as_index = False).median()
med_price_items.rename(columns = {'item_price': 'med_price_all_items'}, inplace = True)

med_price_items.head()

,year,month,med_price_all_items
0,2013,1,299.0
1,2013,2,299.0
2,2013,3,299.0
3,2013,4,299.0
4,2013,5,299.0


#### 2. Weekends and holiday features

In [132]:
# Separate the sales days together
sales_days = sales_full[['date']].copy()
sales_days.drop_duplicates(inplace = True, ignore_index = True)
sales_days.sort_values('date', inplace = True, ignore_index = True)

# Create month, year, and weekend indicator variables
sales_days['year'] = sales_days['date'].dt.year
sales_days['month'] = sales_days['date'].dt.month
sales_days['weekend_indicator'] = sales_days['date'].case_when(
    [(sales_days['date'].dt.day_of_week.isin([5, 6]), 1),
     (~sales_days['date'].dt.day_of_week.isin([5, 6]), 0)]
)

sales_days.head(7)

,date,year,month,weekend_indicator
0,2013-01-01,2013,1,0
1,2013-01-02,2013,1,0
2,2013-01-03,2013,1,0
3,2013-01-04,2013,1,0
4,2013-01-05,2013,1,1
5,2013-01-06,2013,1,1
6,2013-01-07,2013,1,0


In [133]:
# See Analysis File for more details
# List of holidays in Russia: https://www.timeanddate.com/holidays/russia/2013 (same as for 2014 and 2015)
holiday_list_2013 = ['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04', '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
                     '2013-02-23', '2013-03-08', '2013-05-01', '2013-05-02', '2013-05-03', '2013-05-09', '2013-05-10',
                     '2013-06-12', '2013-11-04']

holiday_list_2014 = ['2014-01-01', '2014-01-02', '2014-01-03', '2014-01-06', '2014-01-07', '2014-01-08',
                     '2014-02-22', '2014-02-23', '2014-03-08', '2014-03-09', '2014-03-10', '2014-05-01', '2014-05-02',
                     '2014-05-03', '2014-05-09', '2014-05-10', '2014-05-11', '2014-06-12', '2014-06-13', '2014-06-14', '2014-06-15',
                     '2014-11-01', '2014-11-02', '2014-11-03', '2014-11-04']

holiday_list_2015 = ['2015-01-01', '2015-01-02', '2015-01-03', '2015-01-04', '2015-01-05', '2015-01-06', '2015-01-07',
                     '2015-01-08', '2015-01-09', '2015-02-23', '2015-03-08', '2015-03-09', '2015-05-01', '2015-05-04', '2015-05-09',
                     '2015-05-11', '2015-06-12'] # only count until October 2015

# Combine to have a list of holidays in training data
holidays = holiday_list_2013 + holiday_list_2014 + holiday_list_2015
holidays = [pd.Timestamp(x) for x in holidays] # Convert the whole list to timestamp format

# Create holiday indicator variable
sales_days['holiday_indicator'] = sales_days['date'].case_when(
    [(sales_days['date'].isin(holidays), 1),
     (~sales_days['date'].isin(holidays), 0)]
)

sales_days.head()

,date,year,month,weekend_indicator,holiday_indicator
0,2013-01-01,2013,1,0,1
1,2013-01-02,2013,1,0,1
2,2013-01-03,2013,1,0,1
3,2013-01-04,2013,1,0,1
4,2013-01-05,2013,1,1,1


In [134]:
# Count the number of days as weekends and holidays in a (month, year) pair
sales_days.drop('date', axis = 1, inplace = True)
sales_month = sales_days.groupby(['year', 'month'], as_index = False).sum()
sales_month.rename(columns = {'weekend_indicator': 'weekends_in_month', 'holiday_indicator': 'holidays_in_month'}, inplace = True)
sales_month.head(7)

,year,month,weekends_in_month,holidays_in_month
0,2013,1,8,8
1,2013,2,8,1
2,2013,3,10,1
3,2013,4,8,0
4,2013,5,8,5
5,2013,6,10,1
6,2013,7,8,0


#### 3. Seasonal features

In [135]:
sales_month['season'] = sales_month['month'].case_when(
    [(sales_month['month'].isin([3, 4, 5]), 'spring'),
     (sales_month['month'].isin([6, 7, 8]), 'summer'),
     (sales_month['month'].isin([9, 10, 11]), 'fall'),
     (sales_month['month'].isin([12, 1, 2]), 'winter')]
)

# Because winter is very cold in Russia, I will encode so that winter goes first
sales_month['season'] = pd.Categorical(values = sales_month['season'], ordered = True,
                                       categories = ['winter', 'spring', 'summer', 'fall'])

# Use one-hot encoding to turn the seasons into features
# Drop the first column (winter) to avoid multicollinearity
onehot_season = pd.get_dummies(sales_month['season'], drop_first = True)

# Convert data type of all the one-hot encoded values: True/False to 0/1
onehot_season = onehot_season.map(lambda x: int(x))

# Merge the one-hot encoding into the sales_month feature DataFrame
sales_month = pd.concat([sales_month, onehot_season], axis = 1)

# Drop the categorical 'season' column now that we have 3 one-hot encoded season columns
sales_month.drop('season', axis = 1, inplace = True)

sales_month.head(7)

,year,month,weekends_in_month,holidays_in_month,spring,summer,fall
0,2013,1,8,8,0,0,0
1,2013,2,8,1,0,0,0
2,2013,3,10,1,1,0,0
3,2013,4,8,0,1,0,0
4,2013,5,8,5,1,0,0
5,2013,6,10,1,0,1,0
6,2013,7,8,0,0,1,0


#### 4. Lag features

In [136]:
# Keep the necessary columns and aggregate sales data by month
lag_by_month = sales_full[['date_block_num', 'shop_id', 'item_category_id', 'item_id', 'item_cnt_day']].copy()
lag_by_month = lag_by_month.groupby(['date_block_num', 'shop_id', 'item_category_id', 'item_id'], as_index = False).sum()

# A little bit more formatting: renaming and sorting columns
lag_by_month.rename(columns = {'item_cnt_day': 'item_cnt_month'}, inplace = True)
lag_by_month.sort_values(['date_block_num', 'shop_id', 'item_category_id', 'item_id'])
lag_by_month.head()

,date_block_num,shop_id,item_category_id,item_id,item_cnt_month
0,0,2,2,27,1.0
1,0,2,2,1409,1.0
2,0,2,2,1467,1.0
3,0,2,2,1471,2.0
4,0,2,2,1832,1.0


In [137]:
# Thanks to ChatGPT, I now get to know a resampling method with respect to multiple indices

def resampling_by_date_block(df, grouped_col_name):
    '''The DataFrame should have 3 columns named: date_block_num, a grouped column (item_id/ item_category_id/ shop_id),
    and a value column of your choice (such as revenue) that is always the rightmost column.'''

    # Create a list of all possible (id, date_block_num) pairs (from_product here refers to Cartesian product between 2 sets)
    # Including months that the item/category/shop couldn't sell any items
    full_index = pd.MultiIndex.from_product([df[grouped_col_name].unique(), np.arange(34)], names = [grouped_col_name, 'date_block_num'])

    # Resampling the (id, date_block_num) pairs
    # A bit of a roundabout when I have to convert them to index and then back to normal columns
    df = df.set_index([grouped_col_name, 'date_block_num'], drop = True)
    df = df.reindex(full_index).reset_index()

    # For the pairs not included in the original dataset, I impute with 0 (as in the case of revenue)
    df = df.fillna(0)

    return df

In [138]:
# Create lag features for a column across each item
# https://tichmangono.github.io/tutorials/2018/05/04/time-series-data-munging-lagging-variables-across-multiple-groups

def group_lag_or_rolling(df, grouped_col_name, method = 'lag', periods = 1, new_col_name = 'new_col'):
    '''The DataFrame used as input should have the same expected input format as the resampling_by_date_block function.
    Outputs a new DataFrame.
    The default method is 'lag', but it can also take 'rolling_mean' and 'rolling_sum' as values.
    "periods" should be an integer denoting the number of lagged periods (default = 1).'''

    # Shift by item_id across date_block_num
    new_df = df.set_index(['date_block_num', grouped_col_name])

    if method == 'lag':
        new_df = new_df.unstack().shift(periods = periods)
    elif method == 'rolling_mean':
        new_df = new_df.unstack().rolling(periods).mean()
    elif method == 'rolling_sum':
        new_df = new_df.unstack().rolling(periods).sum()
    else:
        raise ValueError("The method can only be one of three options: 'lag', 'rolling_mean', or 'rolling_sum'.")

    # Restore the unstacked indices to its original order
    new_df = new_df.stack(future_stack = True)
    new_df = new_df.reset_index()

    # Additional formatting: filling NaN values with 0 and change column names
    new_df = new_df.fillna(0)
    new_df = new_df.rename(columns = {new_df.columns[2]: new_col_name})

    return new_df

In [139]:
lag_by_month_item = lag_by_month[['date_block_num', 'item_id', 'item_cnt_month']]\
    .groupby(['date_block_num', 'item_id'], as_index = False).sum()

lag_by_month_item = resampling_by_date_block(lag_by_month_item, 'item_id')
lag_by_month_item.head()

,item_id,date_block_num,item_cnt_month
0,19,0,1.0
1,19,1,0.0
2,19,2,0.0
3,19,3,0.0
4,19,4,0.0


In [140]:
lag_by_month_item_1m = group_lag_or_rolling(lag_by_month_item, 'item_id', periods = 1, new_col_name = 'total_item_cnt_lag_1m')
lag_by_month_item_2m = group_lag_or_rolling(lag_by_month_item, 'item_id', periods = 2, new_col_name = 'total_item_cnt_lag_2m')
lag_by_month_item_3m = group_lag_or_rolling(lag_by_month_item, 'item_id', periods = 3, new_col_name = 'total_item_cnt_lag_3m')

# Example of a lag feature
lag_by_month_item_1m.head()

,date_block_num,item_id,total_item_cnt_lag_1m
0,0,0,0.0
1,0,1,0.0
2,0,2,0.0
3,0,3,0.0
4,0,4,0.0


In [141]:
# Test with item_id = 22167
lag_by_month_item[lag_by_month_item['item_id'] == 22167].tail()

,item_id,date_block_num,item_cnt_month
451923,22167,29,33.0
451924,22167,30,34.0
451925,22167,31,29.0
451926,22167,32,21.0
451927,22167,33,37.0


In [142]:
lag_by_month_item_1m[lag_by_month_item_1m['item_id'] == 22167].tail()

,date_block_num,item_id,total_item_cnt_lag_1m
654207,29,22167,33.0
676014,30,22167,33.0
697821,31,22167,34.0
719628,32,22167,29.0
741435,33,22167,21.0


In [143]:
# Merge into the original "month and item" DataFrame
lag_by_month_item_full = lag_by_month_item.merge(lag_by_month_item_1m, how = 'left', on = ['date_block_num', 'item_id'])\
    .merge(lag_by_month_item_2m, how = 'left', on = ['date_block_num', 'item_id'])\
    .merge(lag_by_month_item_3m, how = 'left', on = ['date_block_num', 'item_id'])

# Drop the 'item_cnt_month' variable in the original table
lag_by_month_item_full.drop('item_cnt_month', axis = 1, inplace = True)
lag_by_month_item_full.head()

,item_id,date_block_num,total_item_cnt_lag_1m,total_item_cnt_lag_2m,total_item_cnt_lag_3m
0,19,0,0.0,0.0,0.0
1,19,1,1.0,0.0,0.0
2,19,2,0.0,1.0,0.0
3,19,3,0.0,0.0,1.0
4,19,4,0.0,0.0,0.0


In [144]:
# Popular item: binary variables that tells which items are sold in the last 3 months
lag_by_month_item_full['highly_popular'] = (lag_by_month_item_full['total_item_cnt_lag_1m'] > 0) & (lag_by_month_item_full['total_item_cnt_lag_2m'] > 0) & (lag_by_month_item_full['total_item_cnt_lag_3m'] > 0)
lag_by_month_item_full['highly_popular'] = lag_by_month_item_full['highly_popular'].astype(int)

lag_by_month_item_full.head()

,item_id,date_block_num,total_item_cnt_lag_1m,total_item_cnt_lag_2m,total_item_cnt_lag_3m,highly_popular
0,19,0,0.0,0.0,0.0,0
1,19,1,1.0,0.0,0.0,0
2,19,2,0.0,1.0,0.0,0
3,19,3,0.0,0.0,1.0,0
4,19,4,0.0,0.0,0.0,0


#### 5. Rolling features

In [145]:
rolling_by_month_item = lag_by_month_item.copy()

# Check with item_id = 22167
rolling_by_month_item[rolling_by_month_item['item_id'] == 22167].tail()

,item_id,date_block_num,item_cnt_month
451923,22167,29,33.0
451924,22167,30,34.0
451925,22167,31,29.0
451926,22167,32,21.0
451927,22167,33,37.0


In [146]:
rolling_mean_3m = group_lag_or_rolling(rolling_by_month_item, 'item_id', method = 'rolling_mean', periods = 3,
                                       new_col_name = 'rolling_mean_3m')

# Check the efficiency of the function
rolling_mean_3m[rolling_mean_3m['item_id'] == 22167].tail()

,date_block_num,item_id,rolling_mean_3m
654207,29,22167,34.666667
676014,30,22167,33.333333
697821,31,22167,32.000000
719628,32,22167,28.000000
741435,33,22167,29.000000


In [147]:
# Lag by 1 month because the current rolling mean will not be in the test set
rolling_mean_3m_lag_1m = rolling_mean_3m.copy()
rolling_mean_3m_lag_1m['date_block_num'] = rolling_mean_3m_lag_1m['date_block_num'] + 1
rolling_mean_3m_lag_1m.rename(columns = {'rolling_mean_3m': 'rolling_mean_3m_lag_1m'}, inplace = True)

rolling_mean_3m_lag_1m[rolling_mean_3m_lag_1m['item_id'] == 22167].tail()

,date_block_num,item_id,rolling_mean_3m_lag_1m
654207,30,22167,34.666667
676014,31,22167,33.333333
697821,32,22167,32.000000
719628,33,22167,28.000000
741435,34,22167,29.000000


#### 6. Add more contextual information

According to Kaggle users, the first word in a shop's name refers to its city, and the first word in an item category refers to a major category.

In [148]:
from sklearn.preprocessing import LabelEncoder

In [149]:
def first_word(str, sep = ' '):
    return str.split(sep = sep)[0]

# Separate the city from shop name
shops['city'] = shops['shop_name'].apply(first_word)
shops['city'] = shops['city'].replace({'!Якутск': 'Якутск'}) # remove the "!" at the beginning of a shop name

le_city = LabelEncoder()
shops['city_id'] = le_city.fit_transform(shops['city']) # encode each city with a number for easier processing

shops.head()

,shop_name,shop_id,city,city_id
0,"!Якутск Орджоникидзе, 56 фран",0,Якутск,29
1,"!Якутск ТЦ ""Центральный"" фран",1,Якутск,29
2,"Адыгея ТЦ ""Мега""",2,Адыгея,0
3,"Балашиха ТРК ""Октябрь-Киномир""",3,Балашиха,1
4,"Волжский ТЦ ""Волга Молл""",4,Волжский,2


In [150]:
item_categories['major_category'] = item_categories['item_category_name'].apply(first_word, sep = ' - ')

le_major_cat = LabelEncoder()
item_categories['major_category_id'] = le_major_cat.fit_transform(item_categories['major_category'])

item_categories.head()

,item_category_name,item_category_id,major_category,major_category_id
0,PC - Гарнитуры/Наушники,0,PC,0
1,Аксессуары - PS2,1,Аксессуары,1
2,Аксессуары - PS3,2,Аксессуары,1
3,Аксессуары - PS4,3,Аксессуары,1
4,Аксессуары - PSP,4,Аксессуары,1


#### 7. Selling status of the item - brand new

In [151]:
# Select the necessary columns and group sales of each (shop_id, item_id) pair by month
sales_by_month = sales_full[['date_block_num', 'shop_id', 'item_id', 'item_cnt_day']].copy()
sales_by_month = sales_by_month.groupby(['date_block_num', 'shop_id', 'item_id'], as_index = False).sum()

sales_by_month.sort_values(['shop_id', 'item_id', 'date_block_num'], inplace = True)
sales_by_month.rename(columns = {'item_cnt_day': 'item_cnt_month'}, inplace = True)

sales_by_month.head()

,date_block_num,shop_id,item_id,item_cnt_month
0,0,2,27,1.0
928366,17,2,27,1.0
123159,2,2,30,1.0
295070,5,2,30,1.0
837965,15,2,30,1.0


In [152]:
brand_new = sales_by_month.drop_duplicates(subset = ['shop_id', 'item_id']).copy()

# brand_new is a binary variable that returns 1 when it is the first month this particular item is sold in this particular shop, 0 otherwise
brand_new.loc[:, 'brand_new'] = 1
brand_new.rename(columns = {'date_block_num': 'first_date_block_sold'}, inplace = True)
brand_new.drop('item_cnt_month', axis = 1, inplace = True)

brand_new.head()

,first_date_block_sold,shop_id,item_id,brand_new
0,0,2,27,1
123159,2,2,30,1
63224,1,2,31,1
687724,12,2,32,1
1,0,2,33,1


#### Combining features:

In [153]:
monthly_features = shops_and_items\
    .merge(cat_and_items, how = 'inner', on = ['year', 'month'])\
    .merge(med_price_items, how = 'inner', on = ['year', 'month'])\
    .merge(sales_month, how = 'inner', on = ['year', 'month'])

monthly_features.head(7)

,year,month,shop_id,med_price_shop,unique_items_in_shop,item_category_id,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall
0,2013,1,2,399.0,728,0,148.00,1,299.0,8,8,0,0,0
1,2013,1,2,399.0,728,1,398.00,2,299.0,8,8,0,0,0
2,2013,1,2,399.0,728,2,1499.00,289,299.0,8,8,0,0,0
3,2013,1,2,399.0,728,3,499.00,2,299.0,8,8,0,0,0
4,2013,1,2,399.0,728,4,599.00,96,299.0,8,8,0,0,0
5,2013,1,2,399.0,728,5,1699.00,74,299.0,8,8,0,0,0
6,2013,1,2,399.0,728,6,1493.75,19,299.0,8,8,0,0,0


In [154]:
final_features = sales_features.merge(monthly_features, how = 'left', on = ['year', 'month', 'shop_id', 'item_category_id'])

final_features.head(7)

,year,month,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall
0,2013,1,0,2,2,27,2499.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0
1,2013,1,0,2,2,1409,1398.5,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0
2,2013,1,0,2,2,1467,899.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0
3,2013,1,0,2,2,1471,2599.0,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0
4,2013,1,0,2,2,1832,1999.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0
5,2013,1,0,2,2,2281,499.5,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0
6,2013,1,0,2,2,2307,2799.0,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0


In [155]:
# Encode month using trigonometric encoding for cyclical features
final_features['month_sin'] = np.sin(final_features['month'] / 12 * 2 * np.pi)
final_features['month_cos'] = np.cos(final_features['month'] / 12 * 2 * np.pi)

# Drop the month variable now that it isn't necessary
final_features.drop('month', axis = 1, inplace = True)

final_features.head(7)

,year,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos
0,2013,0,2,2,27,2499.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025
1,2013,0,2,2,1409,1398.5,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025
2,2013,0,2,2,1467,899.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025
3,2013,0,2,2,1471,2599.0,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025
4,2013,0,2,2,1832,1999.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025
5,2013,0,2,2,2281,499.5,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025
6,2013,0,2,2,2307,2799.0,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025


In [156]:
# Merge other useful feature information
final_features = final_features.merge(lag_by_month_item_full, how = 'left', on = ['date_block_num', 'item_id'])\
    .merge(rolling_mean_3m_lag_1m, how = 'left', on = ['date_block_num', 'item_id'])\
    .merge(shops[['shop_id', 'city_id']], how = 'left', on = 'shop_id')\
    .merge(item_categories[['item_category_id', 'major_category_id']], how = 'left', on = 'item_category_id')\
    .merge(brand_new, how = 'left', on = ['shop_id', 'item_id'])

final_features['rolling_mean_3m_lag_1m'] = final_features['rolling_mean_3m_lag_1m'].fillna(0)

pd.set_option('display.max_columns', None) # display all columns in the DataFrame

# Change some data types
final_features['weekends_in_month'] = final_features['weekends_in_month'].astype('int')
final_features['holidays_in_month'] = final_features['holidays_in_month'].astype('int')

final_features.head(7)

,year,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos,total_item_cnt_lag_1m,total_item_cnt_lag_2m,total_item_cnt_lag_3m,highly_popular,rolling_mean_3m_lag_1m,city_id,major_category_id,first_date_block_sold,brand_new
0,2013,0,2,2,27,2499.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
1,2013,0,2,2,1409,1398.5,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
2,2013,0,2,2,1467,899.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
3,2013,0,2,2,1471,2599.0,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
4,2013,0,2,2,1832,1999.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
5,2013,0,2,2,2281,499.5,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
6,2013,0,2,2,2307,2799.0,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1


In [157]:
final_features.shape

(1096658, 28)

#### Explanation of new feature names: (work in progress)

- ```med_price_shop```, ```unique_items_in_shop```: Median price of an item and number of unique items sold in a particular shop in the month

- ```med_price_cat```, ```unique_items_in_cat```: Median price of an item and number of unique items belonging to a particular category sold in the month

- ```med_price_all_items```: Median price of all items sold in the month

- ```weekends_in_month```: Number of days of weekend (Saturday or Sunday) in the month

- ```holidays_in_month```: Number of days of public holiday in the month

- ```spring```, ```summer```, ```fall```: Dummy variables denoting the season that an item was sold. If ```spring = summer = fall = 0```, then this item was sold in winter

- ```month_sin```, ```month_cos```: Month in a year, encoded using trigonometric encoding to reflect its cyclical nature throughout a year.

According to my personal observation when testing different learning models, the city and major category variable looks very important.

### IV. Declare training and validation sets

We are using 5-fold time series split cross-validation. However, because there are more than 1 observation in a single time frame (panel data), I have to split the folds manually.

Splitting plan (using date_block_num): There are 34 months in the training data (indexed from 0 to 33). I will split the folds so that for the smallest fold, there is at least sufficient data for a year to capture the cyclical characteristics of time series.

| Fold | Fold's training data          | Fold's validation data   |
|------|-------------------------------|--------------------------|
| 1    | ```date_block_num```: 0 -> 11 | ```date_block_num```: 12 |
| 2    | ```date_block_num```: 0 -> 17 | ```date_block_num```: 18 |
| 3    | ```date_block_num```: 0 -> 23 | ```date_block_num```: 24 |
| 4    | ```date_block_num```: 0 -> 29 | ```date_block_num```: 30 |
| 5    | ```date_block_num```: 0 -> 32 | ```date_block_num```: 33 |

In [158]:
# Split into the corresponding folds

# Fold 1
fold_1_train = final_features[final_features['date_block_num'].isin(range(12))]
fold_1_test = final_features[final_features['date_block_num'] == 12]

# Fold 2
fold_2_train = final_features[final_features['date_block_num'].isin(range(18))]
fold_2_test = final_features[final_features['date_block_num'] == 18]

# Fold 3
fold_3_train = final_features[final_features['date_block_num'].isin(range(24))]
fold_3_test = final_features[final_features['date_block_num'] == 24]

# Fold 4
fold_4_train = final_features[final_features['date_block_num'].isin(range(30))]
fold_4_test = final_features[final_features['date_block_num'] == 30]

# Fold 5
fold_5_train = final_features[final_features['date_block_num'].isin(range(33))]
fold_5_test = final_features[final_features['date_block_num'] == 33]

In [159]:
# Number of data points in each fold
print(f'''Fold 1: {fold_1_train.shape[0]} items in training set, {fold_1_test.shape[0]} items in validation set.
      Validation / Training = {fold_1_test.shape[0] / fold_1_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 2: {fold_2_train.shape[0]} items in training set, {fold_2_test.shape[0]} items in validation set.
      Validation / Training = {fold_2_test.shape[0] / fold_2_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 3: {fold_3_train.shape[0]} items in training set, {fold_3_test.shape[0]} items in validation set.
      Validation / Training = {fold_3_test.shape[0] / fold_3_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 4: {fold_4_train.shape[0]} items in training set, {fold_4_test.shape[0]} items in validation set.
      Validation / Training = {fold_4_test.shape[0] / fold_4_train.shape[0]}''')
print('-------------------------------------------------------------')
print(f'''Fold 5: {fold_5_train.shape[0]} items in training set, {fold_5_test.shape[0]} items in validation set.
      Validation / Training = {fold_5_test.shape[0] / fold_5_train.shape[0]}''')

Fold 1: 482291 items in training set, 35056 items in validation set.
      Validation / Training = 0.07268640716911574
-------------------------------------------------------------
Fold 2: 674749 items in training set, 31616 items in validation set.
      Validation / Training = 0.04685594198731677
-------------------------------------------------------------
Fold 3: 856606 items in training set, 29738 items in validation set.
      Validation / Training = 0.03471607716966727
-------------------------------------------------------------
Fold 4: 1007927 items in training set, 23354 items in validation set.
      Validation / Training = 0.02317032880357407
-------------------------------------------------------------
Fold 5: 1074838 items in training set, 21820 items in validation set.
      Validation / Training = 0.02030073369196102


In [160]:
# Split into features and targets

# Fold 1
X_train_1 = fold_1_train.drop('item_cnt_month', axis = 1)#.values
y_train_1 = fold_1_train[['item_cnt_month']].values.ravel()
X_test_1 = fold_1_test.drop('item_cnt_month', axis = 1)#.values
y_test_1 = fold_1_test[['item_cnt_month']].values.ravel()

# Fold 2
X_train_2 = fold_2_train.drop('item_cnt_month', axis = 1)#.values
y_train_2 = fold_2_train[['item_cnt_month']].values.ravel()
X_test_2 = fold_2_test.drop('item_cnt_month', axis = 1)#.values
y_test_2 = fold_2_test[['item_cnt_month']].values.ravel()

# Fold 3
X_train_3 = fold_3_train.drop('item_cnt_month', axis = 1)#.values
y_train_3 = fold_3_train[['item_cnt_month']].values.ravel()
X_test_3 = fold_3_test.drop('item_cnt_month', axis = 1)#.values
y_test_3 = fold_3_test[['item_cnt_month']].values.ravel()

# Fold 4
X_train_4 = fold_4_train.drop('item_cnt_month', axis = 1)#.values
y_train_4 = fold_4_train[['item_cnt_month']].values.ravel()
X_test_4 = fold_4_test.drop('item_cnt_month', axis = 1)#.values
y_test_4 = fold_4_test[['item_cnt_month']].values.ravel()

# Fold 5
X_train_5 = fold_5_train.drop('item_cnt_month', axis = 1)#.values
y_train_5 = fold_5_train[['item_cnt_month']].values.ravel()
X_test_5 = fold_5_test.drop('item_cnt_month', axis = 1)#.values
y_test_5 = fold_5_test[['item_cnt_month']].values.ravel()

In [161]:
# Check the shape of each X and y in training and test sets
print(f'''Fold 1: X_train: {X_train_1.shape}, y_train: {y_train_1.shape}, X_test: {X_test_1.shape}, y_test: {y_test_1.shape}''')
print(f'''Fold 2: X_train: {X_train_2.shape}, y_train: {y_train_2.shape}, X_test: {X_test_2.shape}, y_test: {y_test_2.shape}''')
print(f'''Fold 3: X_train: {X_train_3.shape}, y_train: {y_train_3.shape}, X_test: {X_test_3.shape}, y_test: {y_test_3.shape}''')
print(f'''Fold 4: X_train: {X_train_4.shape}, y_train: {y_train_4.shape}, X_test: {X_test_4.shape}, y_test: {y_test_4.shape}''')
print(f'''Fold 5: X_train: {X_train_5.shape}, y_train: {y_train_5.shape}, X_test: {X_test_5.shape}, y_test: {y_test_5.shape}''')

Fold 1: X_train: (482291, 27), y_train: (482291,), X_test: (35056, 27), y_test: (35056,)
Fold 2: X_train: (674749, 27), y_train: (674749,), X_test: (31616, 27), y_test: (31616,)
Fold 3: X_train: (856606, 27), y_train: (856606,), X_test: (29738, 27), y_test: (29738,)
Fold 4: X_train: (1007927, 27), y_train: (1007927,), X_test: (23354, 27), y_test: (23354,)
Fold 5: X_train: (1074838, 27), y_train: (1074838,), X_test: (21820, 27), y_test: (21820,)


In [162]:
# Example on fold 1
fold_1_train.tail()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos,total_item_cnt_lag_1m,total_item_cnt_lag_2m,total_item_cnt_lag_3m,highly_popular,rolling_mean_3m_lag_1m,city_id,major_category_id,first_date_block_sold,brand_new
482286,2013,11,59,75,3160,1590.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0,4.0,11.0,11.0,1,8.666667,30,15,0,1
482287,2013,11,59,75,4186,1491.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0,36.0,35.0,20.0,1,30.333333,30,15,7,1
482288,2013,11,59,75,4688,1499.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0,35.0,30.0,40.0,1,35.000000,30,15,0,1
482289,2013,11,59,75,12727,1490.0,1.0,299.0,671,1290.0,96,299.0,9,0,0,0,0,-2.449294e-16,1.0,15.0,13.0,23.0,1,17.000000,30,15,1,1
482290,2013,11,59,83,22091,109.0,1.0,299.0,671,79.0,6,299.0,9,0,0,0,0,-2.449294e-16,1.0,16.0,12.0,41.0,1,23.000000,30,19,2,1


In [163]:
# Set seed for reproducibility
my_seed = 275225

In [164]:
categorical_col_indices = ['shop_id', 'item_category_id', 'item_id', 'city_id', 'major_category_id']
minmax_indices = ['year', 'date_block_num', 'med_price_shop', 'unique_items_in_shop', 'med_price_cat', 'unique_items_in_cat',
                                                                  'med_price_all_items', 'weekends_in_month', 'holidays_in_month', 'total_item_cnt_lag_1m',
                                                                  'total_item_cnt_lag_2m', 'total_item_cnt_lag_3m', 'rolling_mean_3m_lag_1m']

In [165]:
from sklearn.preprocessing import TargetEncoder, MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Encode categorical variables (shop_id, item_category_id, item_id) with TargetEncoder
# Encode all other variables (except for "month" and dummy variables) with MinMaxScaler, default range [0, 1]

target_enc = TargetEncoder(target_type = 'continuous', shuffle = True, random_state = my_seed)

transformer = ColumnTransformer(
    transformers = [('target_enc', target_enc, categorical_col_indices),
                    ('minmax_scaler', MinMaxScaler(), minmax_indices)],
    remainder = 'passthrough'
)

### V. Implementing regression models and hyperparameter tuning

In [166]:
from sklearn.metrics import root_mean_squared_error
import statistics
import time # I want to measure the efficiency of different machine learning models

# Custom function for training in many folds and many different models

folds = [[X_train_1, X_test_1, y_train_1, y_test_1], [X_train_2, X_test_2, y_train_2, y_test_2],
         [X_train_3, X_test_3, y_train_3, y_test_3], [X_train_4, X_test_4, y_train_4, y_test_4],
         [X_train_5, X_test_5, y_train_5, y_test_5]]

def kfold_regression(model_name, model, transformer_x, folds, output = 'df'):
    # Choices of output:
    # (1) 'df' (default, output a neatly formatted DataFrame of the results)
    # (2) 'score' (output only the mean RMSE)
    # (3) 'model' (output the model)

    rmse_folds = []

    pipe = Pipeline([('transformer', transformer_x), ('regression', model)])

    # I want to measure the time that Python fits the data and output the scores for all 5 folds
    start_time = time.perf_counter()

    # Each element in "folds" list should have the form [X_train, X_test, y_train, y_test]
    for fold in folds:
        pipe.fit(fold[0], fold[2])
        y_pred = pipe.predict(fold[1])
        rmse_folds.append(root_mean_squared_error(fold[3], y_pred))
    
    end_time = time.perf_counter()

    if output == 'df':
        df = pd.DataFrame({'model_name': [model_name], 'mean_rmse': [statistics.mean(rmse_folds)],
                       'std_rmse': [statistics.stdev(rmse_folds)], 'execution_time': [end_time - start_time]})
        return df
    elif output == 'score':
        return statistics.mean(rmse_folds)
    elif output == 'model':
        return pipe['regression']
    else:
        raise ValueError('The "output" parameter must be \'df\', \'score\', or \'model\'.')

In [167]:
# Import the necessary models
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor, HistGradientBoostingRegressor
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from lightgbm import LGBMRegressor

In [168]:
import re
import warnings
warnings.filterwarnings(action = 'ignore', message = r'^X does not have valid feature names')

In [169]:
kfold_regression('LightGBM', LGBMRegressor(n_jobs = -1, random_state = my_seed, verbose = -1), transformer, folds)
# Last result: 1.458746

,model_name,mean_rmse,std_rmse,execution_time
0,LightGBM,1.458746,1.990337,54.665751


In [170]:
import optuna
from optuna.samplers import TPESampler

In [171]:
# Tuning for LightGBM
# https://medium.com/@amitsinghrajput_92567/understanding-hyperparameters-in-decision-trees-xgboost-and-lightgbm-7b64cfed77f0

def objective_lgbm(trial):
    num_leaves = trial.suggest_int('num_leaves', 5, 100, step = 1)
    max_depth = trial.suggest_int('max_depth', 5, 100, step = 1)
    learning_rate = trial.suggest_float('learning_rate', 10**(-4), 10**(-1), log = True)
    min_split_gain = trial.suggest_float('min_split_gain', 0.05, 0.95, step = 0.05)
    min_child_weight = trial.suggest_float('min_child_weight', 10**(-4), 10**1, log = True)
    min_child_samples = trial.suggest_int('min_child_samples', 10, 10000, log = True)
    subsample = trial.suggest_float('subsample', 0.05, 0.95, step = 0.05)
    subsample_freq = trial.suggest_int('subsample_freq', 10, 1000, log = True)
    colsample_bytree = trial.suggest_float('colsample_bytree', 0.05, 0.95, step = 0.05)
    reg_alpha = trial.suggest_float('reg_alpha', 10**(-3), 10**1, log = True)
    reg_lambda = trial.suggest_float('reg_lambda', 10**(-3), 10**1, log = True)

    # Set n_estimators = 10000, random_state, verbose, and n_jobs to preferred value
    lgbm_tuned = LGBMRegressor(n_estimators = 10000, random_state = my_seed, n_jobs = -1, verbose = -1,
                               num_leaves = num_leaves, max_depth = max_depth, learning_rate = learning_rate,
                               min_split_gain = min_split_gain, min_child_weight = min_child_weight,
                               min_child_samples = min_child_samples, subsample = subsample,
                               subsample_freq = subsample_freq, colsample_bytree = colsample_bytree,
                               reg_alpha = reg_alpha, reg_lambda = reg_lambda)

    try:
        score = kfold_regression('LightGBM', lgbm_tuned, transformer, folds, output = 'score')

        if trial.should_prune():
            raise optuna.TrialPruned()
    
        return score
    
    except Exception:
        pass

    # Use the try-except syntax to make sure the study does not end when a trial fails

In [172]:
# WARNING: THIS CODE BLOCK WILL TAKE A LONG TIME TO RUN

study = optuna.create_study(sampler = TPESampler(seed = my_seed), direction = 'minimize',
    pruner = optuna.pruners.MedianPruner(n_startup_trials = 3, n_warmup_steps = 1, interval_steps = 1, n_min_trials = 1))

# optuna.logging.set_verbosity(optuna.logging.ERROR) # can uncomment this command to avoid excessive logging by Optuna
study.optimize(objective_lgbm, n_trials = 10)

lgbm_best_params = study.best_trial.params

print(f'Trial number {study.best_trial.number} has the best performance, with a score of {study.best_trial.value}.')
print(f'The best parameters are: {study.best_trial.params}')

[I 2025-08-22 09:48:14,814] A new study created in memory with name: no-name-06592cdd-9102-4ee8-a2b8-0dc22a2ea7c1
[I 2025-08-22 09:59:05,799] Trial 0 finished with value: 1.4221620052766029 and parameters: {'num_leaves': 31, 'max_depth': 64, 'learning_rate': 0.023251443377156827, 'min_split_gain': 0.6000000000000001, 'min_child_weight': 0.053160726817981174, 'min_child_samples': 64, 'subsample': 0.8, 'subsample_freq': 189, 'colsample_bytree': 0.95, 'reg_alpha': 0.31549086474120447, 'reg_lambda': 0.08921894140014044}. Best is trial 0 with value: 1.4221620052766029.
[W 2025-08-22 10:03:31,572] Trial 1 failed with parameters: {'num_leaves': 71, 'max_depth': 61, 'learning_rate': 0.01072080074276754, 'min_split_gain': 0.95, 'min_child_weight': 0.00012498403032937637, 'min_child_samples': 2092, 'subsample': 0.8500000000000001, 'subsample_freq': 719, 'colsample_bytree': 0.05, 'reg_alpha': 3.5398408573361486, 'reg_lambda': 0.008814129528701333} because of the following error: KeyboardInterrupt

KeyboardInterrupt: 

In [ ]:
# Save model to pickle file
import pickle
saved_model = LGBMRegressor(n_estimators = 10000, random_state = my_seed, n_jobs = -1, verbose = -1, **lgbm_best_params)
pickle.dump(saved_model, open('tuned_lgbm.sav', 'wb'))

In [ ]:
# Run the tuned model again as a validation
kfold_regression('LightGBM', pickle.load(open('tuned_lgbm.sav', 'rb')), transformer, folds)
# Last result: 1.460556

,model_name,mean_rmse,std_rmse,execution_time
0,LightGBM,1.422162,1.730136,506.48016


In [ ]:
# Tuning for ElasticNet
def objective_elasticnet(trial):
    alpha = trial.suggest_float('alpha', 10**(-3), 10**2, log = True)
    l1_ratio = trial.suggest_float('l1_ratio', 0.05, 0.95, step = 0.05)
    tol = trial.suggest_float('tol', 10**(-5), 10**(-2), log = True)

    # Increase max_iter to 10000 for better results
    elasticnet_tuned = ElasticNet(random_state = my_seed, max_iter = 10000, alpha = alpha,
                                  l1_ratio = l1_ratio, tol = tol)

    try:
        score = kfold_regression('Elastic Net', elasticnet_tuned, transformer, folds, output = 'score')

        if trial.should_prune():
            raise optuna.TrialPruned()
    
        return score
    
    except Exception:
        pass

    # Use the try-except syntax to make sure the study does not end when a trial fails 

In [ ]:
# WARNING: THIS CODE BLOCK WILL TAKE A LONG TIME TO RUN (est. 2 hours)

# https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.MedianPruner.html
# Use MedianPruner as a form of early stopping
study1 = optuna.create_study(sampler = TPESampler(seed = my_seed), direction = 'minimize',
    pruner = optuna.pruners.MedianPruner(n_startup_trials = 3, n_warmup_steps = 1, interval_steps = 1, n_min_trials = 1))

# Not printing out Optuna updates periodically, including failed trials
optuna.logging.set_verbosity(optuna.logging.ERROR)
study1.optimize(objective_elasticnet, n_trials = 500)

# https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.FrozenTrial.html
elasticnet_best_params = study1.best_trial.params

print(f'Trial number {study1.best_trial.number} has the best performance, with a score of {study1.best_trial.value}.')
print(f'The best parameters are: {study1.best_trial.params}')

[I 2025-08-20 16:46:37,363] A new study created in memory with name: no-name-c2ef5e0e-47fd-413c-ae16-5bbbda95c859


Trial number 10 has the best performance, with a score of 1.9979584161243853.
The best parameters are: {'alpha': 60.76621672979539, 'l1_ratio': 0.95, 'tol': 0.00024882590415386364}


In [ ]:
# Save model to pickle file
import pickle
saved_model = ElasticNet(random_state = my_seed, max_iter = 10000, **elasticnet_best_params)
pickle.dump(saved_model, open('tuned_elasticnet.sav', 'wb'))

In [ ]:
# Run the tuned model again as a validation
kfold_regression('ElasticNet (Tuned)', pickle.load(open('tuned_elasticnet.sav', 'rb')), transformer, folds)

,model_name,mean_rmse,std_rmse,execution_time
0,ElasticNet (Tuned),1.997958,2.523672,6.595186


### VIII. Prepare data for final submission

In [275]:
# Declare final model for submission
from sklearn.ensemble import StackingRegressor
import pickle

final_model = pickle.load(open('tuned_lgbm.sav', 'rb')) # open existing tuned LightGBM model

In [276]:
test_shop_items = test[['shop_id', 'item_id']].copy()
test_shop_items.head() # this is the test data stripped of its ID

,shop_id,item_id
0,5,5037
1,5,5320
2,5,5233
3,5,5232
4,5,5268


In [277]:
# Set constant data for some columns
test_features = test_shop_items.copy()

test_features['year'] = 2013
test_features['date_block_num'] = 34 # number of months elapsed since Jan 2013

test_features['spring'] = 0; test_features['summer'] = 0; test_features['fall'] = 1 # November is in the fall season
test_features['weekends_in_month'] = 9
test_features['holidays_in_month'] = 1 # Russia's Unity Day in 4 November

# For the month_sin and month_cos trigonometric encoding, replace the month number with 11
test_features['month_sin'] = np.sin(11 / 12 * 2 * np.pi)
test_features['month_cos'] = np.cos(11 / 12 * 2 * np.pi)

test_features.head()

,shop_id,item_id,year,date_block_num,spring,summer,fall,weekends_in_month,holidays_in_month,month_sin,month_cos
0,5,5037,2013,34,0,0,1,9,1,-0.5,0.866025
1,5,5320,2013,34,0,0,1,9,1,-0.5,0.866025
2,5,5233,2013,34,0,0,1,9,1,-0.5,0.866025
3,5,5232,2013,34,0,0,1,9,1,-0.5,0.866025
4,5,5268,2013,34,0,0,1,9,1,-0.5,0.866025


In [278]:
# Retrieve item category and item price (each item only belongs to 1 category and has 1 price)
item_info = final_features[['shop_id', 'item_id', 'item_category_id', 'item_price']].copy()

# Keep only the last occurence of (shop_id, item_id) pair (showing the latest price)
item_info.drop_duplicates(subset = ['shop_id', 'item_id'], keep = 'last', inplace = True, ignore_index = True)
item_info.head()

,shop_id,item_id,item_category_id,item_price
0,2,3056,2,899.0
1,2,4464,2,599.0
2,2,5649,2,2190.0
3,2,6705,2,2499.0
4,2,11392,2,899.0


In [279]:
# Merge item information to the test set
item_info = test_shop_items.merge(item_info, on = ['shop_id', 'item_id'], how = 'left')
item_info.drop_duplicates(inplace = True, ignore_index = True)
item_info.head()

,shop_id,item_id,item_category_id,item_price
0,5,5037,2.0,749.5
1,5,5320,NaN,NaN
2,5,5233,2.0,1199.0
3,5,5232,23.0,599.0
4,5,5268,NaN,NaN


In [280]:
item_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214200 entries, 0 to 214199
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   shop_id           214200 non-null  int64  
 1   item_id           214200 non-null  int64  
 2   item_category_id  102334 non-null  float64
 3   item_price        102334 non-null  float64
dtypes: float64(2), int64(2)
memory usage: 6.5 MB


In [281]:
# Create a new item category ID that denotes "unknown"
new_cat_id = item_categories['item_category_id'].max() + 1 # new identifier for unknown category
print(new_cat_id)

84


In [282]:
# item_info['item_category_id']
item_info['item_category_id'] = item_info['item_category_id'].fillna(new_cat_id)
item_info.head()

,shop_id,item_id,item_category_id,item_price
0,5,5037,2.0,749.5
1,5,5320,84.0,NaN
2,5,5233,2.0,1199.0
3,5,5232,23.0,599.0
4,5,5268,84.0,NaN


In [283]:
from sklearn.impute import SimpleImputer
imp = SimpleImputer(strategy = 'median')

imputed_data = imp.fit_transform(item_info) # outputs a NumPy array
new_item_info = pd.DataFrame(imputed_data, columns = item_info.columns) # convert into a DataFrame
new_item_info.head()

,shop_id,item_id,item_category_id,item_price
0,5.0,5037.0,2.0,749.5
1,5.0,5320.0,84.0,449.0
2,5.0,5233.0,2.0,1199.0
3,5.0,5232.0,23.0,599.0
4,5.0,5268.0,84.0,449.0


In [284]:
# Create new features based on imputed data (for shops)
new_item_info_shop = new_item_info[['shop_id', 'item_id', 'item_price']].copy()
new_item_info_shop = new_item_info_shop.groupby('shop_id', as_index = False).agg({'item_price': 'median', 'item_id': 'nunique'})
new_item_info_shop.rename(columns = {'item_price': 'med_price_shop', 'item_id': 'unique_items_in_shop'}, inplace = True)
new_item_info_shop.head()

,shop_id,med_price_shop,unique_items_in_shop
0,2.0,449.0,5100
1,3.0,449.0,5100
2,4.0,449.0,5100
3,5.0,449.0,5100
4,6.0,449.0,5100


In [285]:
# Create new features based on imputed data (for categories)
new_item_info_cat = new_item_info[['item_category_id', 'item_id', 'item_price']].copy()
new_item_info_cat = new_item_info_cat.groupby('item_category_id', as_index = False).agg({'item_price': 'median', 'item_id': 'nunique'})
new_item_info_cat.rename(columns = {'item_price': 'med_price_cat', 'item_id': 'unique_items_in_cat'}, inplace = True)
new_item_info_cat.head()

,item_category_id,med_price_cat,unique_items_in_cat
0,2.0,1199.0,148
1,3.0,2299.0,153
2,4.0,449.0,9
3,5.0,1099.0,29
4,6.0,2290.0,9


In [286]:
# Get median price of all items (including imputed) sold in November 2015 - test data 
new_item_info['med_price_all_items'] = new_item_info['item_price'].median()
new_item_info.head()

,shop_id,item_id,item_category_id,item_price,med_price_all_items
0,5.0,5037.0,2.0,749.5,449.0
1,5.0,5320.0,84.0,449.0,449.0
2,5.0,5233.0,2.0,1199.0,449.0
3,5.0,5232.0,23.0,599.0,449.0
4,5.0,5268.0,84.0,449.0,449.0


In [287]:
# Merge into the existing new_item_info DataFrame
item_info = new_item_info.merge(new_item_info_shop, how = 'left', on = 'shop_id')\
    .merge(new_item_info_cat, how = 'left', on = 'item_category_id')

item_info.head()

,shop_id,item_id,item_category_id,item_price,med_price_all_items,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat
0,5.0,5037.0,2.0,749.5,449.0,449.0,5100,1199.0,148
1,5.0,5320.0,84.0,449.0,449.0,449.0,5100,449.0,5100
2,5.0,5233.0,2.0,1199.0,449.0,449.0,5100,1199.0,148
3,5.0,5232.0,23.0,599.0,449.0,449.0,5100,1199.0,143
4,5.0,5268.0,84.0,449.0,449.0,449.0,5100,449.0,5100


In [288]:
# Find the lag total item count of each item in 1m, 2m, 3m
train_lag_item = final_features[['date_block_num', 'item_id', 'item_cnt_month']][final_features['date_block_num'].isin([31, 32, 33])]
train_lag_item = train_lag_item.groupby(['date_block_num', 'item_id'], as_index = False).sum()

# Reshape the data from long to wide format
train_lag_item = train_lag_item.pivot(index = 'item_id', columns = 'date_block_num', values = 'item_cnt_month')
train_lag_item.reset_index(inplace = True)

# For compatibility and aesthetic reasons, I should rename columns and fill NaN values
train_lag_item.rename_axis(None, axis = 1, inplace = True)
train_lag_item.rename(columns = {31: 'total_item_cnt_lag_3m', 32: 'total_item_cnt_lag_2m', 33: 'total_item_cnt_lag_1m'}, inplace = True)
train_lag_item.fillna(0, inplace = True)

train_lag_item.head()

,item_id,total_item_cnt_lag_3m,total_item_cnt_lag_2m,total_item_cnt_lag_1m
0,30,6.0,3.0,1.0
1,31,12.0,9.0,10.0
2,32,13.0,8.0,10.0
3,33,10.0,11.0,10.0
4,38,5.0,2.0,0.0


In [289]:
# Now merge to create the full test set
test_features = test_features.merge(item_info, how = 'left', on = ['shop_id', 'item_id'])
test_features.head()

,shop_id,item_id,year,date_block_num,spring,summer,fall,weekends_in_month,holidays_in_month,month_sin,month_cos,item_category_id,item_price,med_price_all_items,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat
0,5,5037,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,749.5,449.0,449.0,5100,1199.0,148
1,5,5320,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100
2,5,5233,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,1199.0,449.0,449.0,5100,1199.0,148
3,5,5232,2013,34,0,0,1,9,1,-0.5,0.866025,23.0,599.0,449.0,449.0,5100,1199.0,143
4,5,5268,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100


In [290]:
test_features = test_features.merge(train_lag_item, how = 'left', on = 'item_id')
test_features.fillna(0, inplace = True)
test_features.head()

,shop_id,item_id,year,date_block_num,spring,summer,fall,weekends_in_month,holidays_in_month,month_sin,month_cos,item_category_id,item_price,med_price_all_items,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,total_item_cnt_lag_3m,total_item_cnt_lag_2m,total_item_cnt_lag_1m
0,5,5037,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,749.5,449.0,449.0,5100,1199.0,148,17.0,14.0,11.0
1,5,5320,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100,0.0,0.0,0.0
2,5,5233,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,1199.0,449.0,449.0,5100,1199.0,148,18.0,19.0,16.0
3,5,5232,2013,34,0,0,1,9,1,-0.5,0.866025,23.0,599.0,449.0,449.0,5100,1199.0,143,17.0,13.0,11.0
4,5,5268,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100,0.0,0.0,0.0


Additional features:

In [291]:
# "highly_popular" feature
test_features['highly_popular'] = (test_features['total_item_cnt_lag_1m'] > 0) & (test_features['total_item_cnt_lag_2m'] > 0) & (test_features['total_item_cnt_lag_3m'] > 0)
test_features['highly_popular'] = test_features['highly_popular'].astype(int)

# "brand_new" feature
test_features['brand_new'] = test_features['item_category_id'].case_when([
    (test_features['item_category_id'] != 84, 1),
    (test_features['item_category_id'] == 84, 0)
])

test_features.head()

,shop_id,item_id,year,date_block_num,spring,summer,fall,weekends_in_month,holidays_in_month,month_sin,month_cos,item_category_id,item_price,med_price_all_items,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,total_item_cnt_lag_3m,total_item_cnt_lag_2m,total_item_cnt_lag_1m,highly_popular,brand_new
0,5,5037,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,749.5,449.0,449.0,5100,1199.0,148,17.0,14.0,11.0,1,1.0
1,5,5320,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100,0.0,0.0,0.0,0,0.0
2,5,5233,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,1199.0,449.0,449.0,5100,1199.0,148,18.0,19.0,16.0,1,1.0
3,5,5232,2013,34,0,0,1,9,1,-0.5,0.866025,23.0,599.0,449.0,449.0,5100,1199.0,143,17.0,13.0,11.0,1,1.0
4,5,5268,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100,0.0,0.0,0.0,0,0.0


In [292]:
# "rolling_mean_3m_lag_1m" feature
latest_3m = lag_by_month.copy()
latest_3m = latest_3m[['date_block_num', 'item_id', 'item_cnt_month']].groupby(['date_block_num', 'item_id'], as_index = False).sum()
latest_3m = latest_3m[latest_3m['date_block_num'].isin([31, 32, 33])]

# Validate on 3 different highly-sold items
latest_3m[latest_3m['item_id'].isin([22164, 22166, 22167])]

,date_block_num,item_id,item_cnt_month
223411,31,22164,20.0
223412,31,22166,11.0
223413,31,22167,29.0
228496,32,22164,9.0
228497,32,22166,5.0
228498,32,22167,21.0
233909,33,22164,15.0
233910,33,22166,11.0
233911,33,22167,37.0


In [293]:
latest_3m.head()

,date_block_num,item_id,item_cnt_month
218306,31,30,6.0
218307,31,31,53.0
218308,31,32,30.0
218309,31,33,14.0
218310,31,38,7.0


In [294]:
# Create a list of all possible (id, date_block_num) pairs (from_product here refers to Cartesian product between 2 sets)
# Including months that the item/category/shop couldn't sell any items
latest_3m_index = pd.MultiIndex.from_product([latest_3m['item_id'].unique(), np.arange(31, 34)], names = ['item_id', 'date_block_num'])

# Resampling the (id, date_block_num) pairs
# A bit of a roundabout when I have to convert them to index and then back to normal columns
latest_3m = latest_3m.set_index(['item_id', 'date_block_num'], drop = True)
latest_3m = latest_3m.reindex(latest_3m_index).reset_index()

# For the pairs not included in the original dataset, I impute with 0 (as in the case of revenue)
latest_3m = latest_3m.fillna(0)

latest_3m.head()

,item_id,date_block_num,item_cnt_month
0,30,31,6.0
1,30,32,3.0
2,30,33,1.0
3,31,31,53.0
4,31,32,9.0


In [295]:
latest_3m[latest_3m['item_id'].isin([22164, 22166, 22167])]

,item_id,date_block_num,item_cnt_month
15315,22164,31,20.0
15316,22164,32,9.0
15317,22164,33,15.0
15318,22166,31,11.0
15319,22166,32,5.0
15320,22166,33,11.0
15321,22167,31,29.0
15322,22167,32,21.0
15323,22167,33,37.0


In [296]:
mean_latest_3m = latest_3m[['item_id', 'item_cnt_month']].groupby('item_id', as_index = False).mean()
mean_latest_3m.rename(columns = {'item_cnt_month': 'rolling_mean_3m_lag_1m'}, inplace = True)
mean_latest_3m.tail()

,item_id,rolling_mean_3m_lag_1m
7181,22162,10.333333
7182,22163,25.000000
7183,22164,14.666667
7184,22166,9.000000
7185,22167,29.000000


In [ ]:
test_features = test_features.merge(mean_latest_3m, how = 'left', on = 'item_id')\
    .merge(shops[['shop_id', 'city_id']], how = 'left', on = 'shop_id')\
    .merge(item_categories[['item_category_id', 'major_category_id']], how = 'left', on = 'item_category_id')\
    .merge(brand_new[['shop_id', 'item_id', 'first_date_block_sold']], how = 'left', on = ['shop_id', 'item_id'])

test_features['rolling_mean_3m_lag_1m'] = test_features['rolling_mean_3m_lag_1m'].fillna(0)
test_features['major_category_id'] = test_features['major_category_id'].fillna(item_categories['major_category_id'].max() + 1)
test_features['first_date_block_sold'] = test_features['first_date_block_sold'].fillna(34)
# All shops mentioned in the test set have appeared in the training set before, so there is no need to fill null city_id values

test_features.head()

,shop_id,item_id,year,date_block_num,spring,summer,fall,weekends_in_month,holidays_in_month,month_sin,month_cos,item_category_id,item_price,med_price_all_items,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,total_item_cnt_lag_3m,total_item_cnt_lag_2m,total_item_cnt_lag_1m,highly_popular,brand_new,rolling_mean_3m_lag_1m,city_id,major_category_id,first_date_block_sold
0,5,5037,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,749.5,449.0,449.0,5100,1199.0,148,17.0,14.0,11.0,1,1.0,85.333333,3,1.0,20.0
1,5,5320,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100,0.0,0.0,0.0,0,0.0,0.000000,3,20.0,34.0
2,5,5233,2013,34,0,0,1,9,1,-0.5,0.866025,2.0,1199.0,449.0,449.0,5100,1199.0,148,18.0,19.0,16.0,1,1.0,90.666667,3,1.0,28.0
3,5,5232,2013,34,0,0,1,9,1,-0.5,0.866025,23.0,599.0,449.0,449.0,5100,1199.0,143,17.0,13.0,11.0,1,1.0,47.666667,3,5.0,31.0
4,5,5268,2013,34,0,0,1,9,1,-0.5,0.866025,84.0,449.0,449.0,449.0,5100,449.0,5100,0.0,0.0,0.0,0,0.0,0.000000,3,20.0,34.0


In [298]:
# Create submission training and test data
submission_train = final_features.copy()

# After checking data types (truncated), I need to change data type of the weekends_in_month and holidays_in_month columns from object to int
submission_train['weekends_in_month'] = submission_train['weekends_in_month'].astype('int')
submission_train['holidays_in_month'] = submission_train['weekends_in_month'].astype('int')

submission_train.head()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,item_cnt_month,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos,total_item_cnt_lag_1m,total_item_cnt_lag_2m,total_item_cnt_lag_3m,highly_popular,rolling_mean_3m_lag_1m,city_id,major_category_id,first_date_block_sold,brand_new
0,2013,0,2,2,27,2499.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
1,2013,0,2,2,1409,1398.5,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
2,2013,0,2,2,1467,899.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
3,2013,0,2,2,1471,2599.0,2.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1
4,2013,0,2,2,1832,1999.0,1.0,399.0,728,1499.0,289,299.0,8,8,0,0,0,0.5,0.866025,0.0,0.0,0.0,0,0.0,0,1,0,1


In [299]:
# Reorder the columns to align with the training data
submission_test = test_features.loc[:, final_features.drop('item_cnt_month', axis = 1).columns.tolist()]
submission_test.head()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos,total_item_cnt_lag_1m,total_item_cnt_lag_2m,total_item_cnt_lag_3m,highly_popular,rolling_mean_3m_lag_1m,city_id,major_category_id,first_date_block_sold,brand_new
0,2013,34,5,2.0,5037,749.5,449.0,5100,1199.0,148,449.0,9,1,0,0,1,-0.5,0.866025,11.0,14.0,17.0,1,85.333333,3,1.0,20.0,1.0
1,2013,34,5,84.0,5320,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025,0.0,0.0,0.0,0,0.000000,3,20.0,34.0,0.0
2,2013,34,5,2.0,5233,1199.0,449.0,5100,1199.0,148,449.0,9,1,0,0,1,-0.5,0.866025,16.0,19.0,18.0,1,90.666667,3,1.0,28.0,1.0
3,2013,34,5,23.0,5232,599.0,449.0,5100,1199.0,143,449.0,9,1,0,0,1,-0.5,0.866025,11.0,13.0,17.0,1,47.666667,3,5.0,31.0,1.0
4,2013,34,5,84.0,5268,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025,0.0,0.0,0.0,0,0.000000,3,20.0,34.0,0.0


Do the train-test split and implementing a pipeline process to submit.

In [300]:
# Full pipeline
X_train_submission = submission_train.drop('item_cnt_month', axis = 1)#.values
y_train_submission = submission_train[['item_cnt_month']].values.ravel()
X_test_submission = submission_test#.values

full_pipe = Pipeline([('target_enc', transformer), ('regression', final_model)])

full_pipe.fit(X_train_submission, y_train_submission)

submission_test['item_cnt_month_unadjusted'] = full_pipe.predict(X_test_submission)

# Cap the predicted values in the [0, 20] range
submission_test['item_cnt_month'] = submission_test['item_cnt_month_unadjusted'].case_when\
    ([(submission_test['item_cnt_month_unadjusted'] < 0, 0),
      (submission_test['item_cnt_month_unadjusted'] > 20, 20),
      (np.logical_and(submission_test['item_cnt_month_unadjusted'] >= 0, submission_test['item_cnt_month_unadjusted'] <= 20),
       submission_test['item_cnt_month_unadjusted'])])

submission_test.head()

,year,date_block_num,shop_id,item_category_id,item_id,item_price,med_price_shop,unique_items_in_shop,med_price_cat,unique_items_in_cat,med_price_all_items,weekends_in_month,holidays_in_month,spring,summer,fall,month_sin,month_cos,total_item_cnt_lag_1m,total_item_cnt_lag_2m,total_item_cnt_lag_3m,highly_popular,rolling_mean_3m_lag_1m,city_id,major_category_id,first_date_block_sold,brand_new,item_cnt_month_unadjusted,item_cnt_month
0,2013,34,5,2.0,5037,749.5,449.0,5100,1199.0,148,449.0,9,1,0,0,1,-0.5,0.866025,11.0,14.0,17.0,1,85.333333,3,1.0,20.0,1.0,0.947308,0.947308
1,2013,34,5,84.0,5320,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025,0.0,0.0,0.0,0,0.000000,3,20.0,34.0,0.0,0.920525,0.920525
2,2013,34,5,2.0,5233,1199.0,449.0,5100,1199.0,148,449.0,9,1,0,0,1,-0.5,0.866025,16.0,19.0,18.0,1,90.666667,3,1.0,28.0,1.0,2.407562,2.407562
3,2013,34,5,23.0,5232,599.0,449.0,5100,1199.0,143,449.0,9,1,0,0,1,-0.5,0.866025,11.0,13.0,17.0,1,47.666667,3,5.0,31.0,1.0,1.911381,1.911381
4,2013,34,5,84.0,5268,449.0,449.0,5100,449.0,5100,449.0,9,1,0,0,1,-0.5,0.866025,0.0,0.0,0.0,0,0.000000,3,20.0,34.0,0.0,0.920525,0.920525


In [301]:
# Extract only the necessary columns
submission_without_id = submission_test[['shop_id', 'item_id', 'item_cnt_month']].copy()

# Merge the ID column, then remove the shop_id and item_id column to fit with sample submission
submission = submission_without_id.merge(test, on = ['shop_id', 'item_id'])
submission = submission.loc[:, ['ID', 'item_cnt_month']]
print(submission.shape)
submission.head()

(214200, 2)


,ID,item_cnt_month
0,0,0.947308
1,1,0.920525
2,2,2.407562
3,3,1.911381
4,4,0.920525


In [ ]:
# What if we round the predicted item count to the nearest integer? (uncomment the next 2 lines)
#submission['item_cnt_month'] = submission['item_cnt_month'].round(0)
#submission.head()

,ID,item_cnt_month
0,0,1.0
1,1,1.0
2,2,2.0
3,3,2.0
4,4,1.0


In [304]:
# Export submission to .csv file and send to Kaggle
submission.to_csv('thai_an_le_submission.csv', index = False)

### Useful resources and references:

- https://www.kaggle.com/code/thnhnguyntrngtt/predict-future-sales-0-86-solution (partly written in Vietnamese - LGBM with early stopping)

- https://www.kaggle.com/code/cngnguynt04/lightgbm-0-87442 (contains many useful tips to optimize performance)